src/System.Text.Json/src/System/Text/Json/JsonHelpers.Date.cs

```diff
@@ -24,6 +24,62 @@ private struct DateTimeParseData
             public byte OffsetToken;
         }
 
+        public static string FormatDateTimeOffset(DateTimeOffset value)
+        {
+            Span<byte> span = stackalloc byte[JsonConstants.MaximumFormatDateTimeOffsetLength];
+
+            JsonWriterHelper.WriteDateTimeOffsetTrimmed(span, value, out int bytesWritten);
+
+            return JsonReaderHelper.GetTextFromUtf8(span.Slice(0, bytesWritten));
+        }
+
+        public static string FormatDateTime(DateTime value)
+        {
+            Span<byte> span = stackalloc byte[JsonConstants.MaximumFormatDateTimeOffsetLength];
+
+            JsonWriterHelper.WriteDateTimeTrimmed(span, value, out int bytesWritten);
+
+            return JsonReaderHelper.GetTextFromUtf8(span.Slice(0, bytesWritten));
+        }
+
+        public static bool TryParseAsISO(ReadOnlySpan<char> source, out DateTime value)
+        {
+            if (!JsonReaderHelper.IsValidDateTimeOffsetParseLength(source.Length))
+            {
+                value = default;
+                return false;
+            }
+
+            int length = JsonReaderHelper.GetUtf8ByteCount(source);
+
+            Span<byte> bytes = length <= JsonConstants.StackallocThreshold
+                ? stackalloc byte[JsonConstants.StackallocThreshold]
+                : new byte[length];
+
+            JsonReaderHelper.GetUtf8FromText(source, bytes);
+
+            return TryParseAsISO(bytes.Slice(0, length), out value);
+        }
+
+        public static bool TryParseAsISO(ReadOnlySpan<char> source, out DateTimeOffset value)
+        {
+            if (!JsonReaderHelper.IsValidDateTimeOffsetParseLength(source.Length))
```

Comment:
When this method is called from `TryParseAsISO(ReadOnlySpan<char> source, out DateTime value)`, this check is redundant. Any way to extract out the duplicate logic into a common workhorse and avoid this check?

Reply:
Why is this check redundant?

Reply:
I might be wrong (since I am just reading the code), but the other, DateTime overload looks like its calling into this one. The `source` being passed-in has already been validated there to be in range. I get other call-sites that directly call the `DateTimeOffset` overload would still need this check, which is why I suggesting maybe we can extract out the common code into a helper that both DateTime and DateTimeOffset overloads can.

Reply:
No, each of the `TryParseAsISO(ReadOnlySpan<char>, out DateTime[Offset])` methods call into the `TryParseAsISO(ReadOnlySpan<byte>, out DateTime[Offset])` (notice `char` vs `byte` spans). The `char` methods validate the length, while the `byte` overloads doesn't. There's no "crossing the streams" between `DateTime` and `DateTimeOffset` in here.

Reply:
Ah, yes, thanks for clarifying! That was it (char -> utf8 byte overload).